// @getbrevo/brevo/dist/esm/api/resources/emailCampaigns/types/GetEmailCampaignsResponse.mjs
// This file was auto-generated by Fern from our API Definition.
export var GetEmailCampaignsResponse;
(function (GetEmailCampaignsResponse) {
    let Campaigns;
    (function (Campaigns) {
        let Item;
        (function (Item) {
            /** Status of the campaign */
            Item.Status = {
                Draft: "draft",
                Sent: "sent",
                Archive: "archive",
                Queued: "queued",
                Suspended: "suspended",
                InProcess: "in_process",
                InReview: "in_review",
                Cancelling: "cancelling",
                Cancelled: "cancelled",
            };
            /** Type of campaign */
            Item.Type = {
                Classic: "classic",
                Trigger: "trigger",
            };
        })(Item = Campaigns.Item || (Campaigns.Item = {}));
    })(Campaigns = GetEmailCampaignsResponse.Campaigns || (GetEmailCampaignsResponse.Campaigns = {}));
})(GetEmailCampaignsResponse || (GetEmailCampaignsResponse = {}));
